File: app/src/main/java/in/smartie/quotedesk/data/mapping/Keys.kt
package `in`.smartie.quotedesk.data.mapping

/**
 * Document-id and logical-key handling.
 *
 * Stock is keyed by the PWA's logical product key (`group|model`), which can
 * contain `/` — `hwWheel|SIEBAL58H/V` is a real one. A `/` in a document id
 * builds an invalid Firestore reference, which is audit issue C6, so the same
 * sanitisation has to be used on every read and every write.
 */
object Keys {

    /** Characters Firestore refuses, or that break a document path. */
    fun sanitiseDocId(raw: String): String {
        val cleaned = raw.trim()
            .replace('/', '_')
            .replace("__proto__", "proto")
        return when {
            cleaned.isEmpty() -> "_"
            cleaned == "." -> "_"
            cleaned == ".." -> "__"
            else -> cleaned
        }
    }

    /** The PWA's logical product key, stored as `id`/`key` on every document. */
    fun productKey(group: String, seedModel: String): String = "$group|$seedModel"

    /**
     * The `/stock` document id for a logical key.
     *
     * **A different scheme from [productDocId], and not interchangeable with
     * it.** Confirmed against the approved V8C4 source: a stock document is
     * addressed by the logical key `group|model` with **only `/` replaced**,
     * where a product document is `group__model` with `/ . # $ [ ]` replaced.
     * `gate|SIE2.5MSMALL` is `gate|SIE2.5MSMALL` as a stock id and
     * `gate__SIE2_5MSMALL` as a product id — the dot survives in one and not
     * the other. Using the product function for a stock document would read
     * and write the wrong row.
     *
     * Deliberately nothing else is replaced, because the ids already in
     * Firestore are the PWA's, not the ones Firestore would merely tolerate.
     * A blank key is the caller's to refuse before it gets here; `StockEntry`
     * validation does that.
     */
    fun stockDocId(key: String): String = key.replace('/', '_')

    /**
     * The `/stockMoves` document id: the movement's own `id` field.
     *
     * The v9 rules require `request.resource.data.id == id`, so the two can
     * never drift. Generate it with `generateId("mv_")` **once, before the
     * transaction**, and reuse it if Firestore replays the transaction body.
     */
    fun stockMoveDocId(movementId: String): String = movementId

    /**
     * The characters the PWA's own `docId` replaces (`index.html:5723`),
     * mirrored in `tools/catalogue-import/lib/keys.mjs`: the path separator,
     * and the four Realtime-Database-era characters the PWA has always
     * stripped. Firestore itself only refuses `/`, but the id that matters is
     * the one already on disk, not the one Firestore would tolerate.
     */
    private val PWA_DOC_ID_CHARACTERS = Regex("[/.#\$\\[\\]]")

    /**
     * Canonical product document id for schema v2: one document per product.
     *
     * Seeding wrote `group|model` and editing wrote `group__model`, so the
     * same product could exist twice with the same `id` field — the duplicate
     * Compose key that crashes the beta Products screen (audit C1/P3).
     *
     * This must be **character for character** what the staging importer
     * wrote, or a native write lands on a second document and re-creates the
     * duplicate the canonical id exists to prevent. It previously replaced
     * only `/`, which is right for `SIEBAL58H/V` and wrong for every model
     * carrying a `.`, `#`, `$`, `[` or `]`. Only the model is canonicalised,
     * not the group, because that is what the PWA does.
     *
     * `sanitiseDocId` is deliberately not used here and is deliberately left
     * alone: it guards arbitrary ids, including stock keys, whose own rule is
     * still an open question against the V8C4 source.
     */
    fun productDocId(group: String, seedModel: String): String {
        val canonical = seedModel.replace(PWA_DOC_ID_CHARACTERS, "_")
        return "${group}__$canonical"
    }

    /** Splits either id scheme back into group and model. */
    fun splitProductKey(raw: String): Pair<String, String>? {
        val pipe = raw.indexOf('|')
        if (pipe > 0) return raw.substring(0, pipe) to raw.substring(pipe + 1)
        val underscores = raw.indexOf("__")
        if (underscores > 0) return raw.substring(0, underscores) to raw.substring(underscores + 2)
        return null
    }

    fun isLegacyProductDocId(docId: String): Boolean = docId.contains('|')

    /** Emails identify a person across duplicate uid records; compare folded. */
    fun normaliseEmail(raw: String?): String = raw?.trim()?.lowercase().orEmpty()

    private const val ID_ALPHABET = "abcdefghijklmnopqrstuvwxyz0123456789"

    /** PWA-compatible ids: `ta_…`, `mv_…`, `c_…`, `q_…`. */
    fun generateId(prefix: String, now: Long = System.currentTimeMillis()): String {
        val time = now.toString(36)
        val random = (1..5).map { ID_ALPHABET.random() }.joinToString("")
        return "$prefix$time$random"
    }
}

File: tools/catalogue-import/lib/keys.mjs
/**
 * Product identity, exactly as the PWA computes it.
 *
 * `docId` (`index.html:5723`) replaces `/ . # $ [ ]`, and thirteen of the 403
 * models need it. The native `Keys.productDocId` replaces the same six —
 * it once replaced only `/`, which is right for `SIEBAL58H/V` and wrong for
 * every model carrying a `.`, and this comment described that older state
 * until N5.7. The two are now character for character the same function, and
 * they have to stay that way: a native write computing a different id lands
 * on a second document and re-creates the duplicate the canonical id exists
 * to prevent.
 */

/** `index.html:5682` — the logical key stored as `id` and `key`. */
export const productKey = (group, seedModel) => `${group}|${seedModel}`;

/** `index.html:5723` — the canonical schema-v2 document id. */
export const productDocId = (group, seedModel) =>
  `${group}__${String(seedModel).replace(/[/.#$[\]]/g, '_')}`;

/** `index.html:5681` — letters and digits only, for duplicate detection. */
export const normalise = (value) =>
  String(value ?? '').toLowerCase().replace(/[^a-z0-9]/g, '');

/** Splits either document-id scheme back into group and model. */
export function splitProductKey(raw) {
  const pipe = String(raw).indexOf('|');
  if (pipe > 0) return [raw.slice(0, pipe), raw.slice(pipe + 1)];
  const underscores = String(raw).indexOf('__');
  if (underscores > 0) return [raw.slice(0, underscores), raw.slice(underscores + 2)];
  return null;
}

/** Models whose document id needs a character replaced; expected to be 13. */
export const needsSanitising = (seedModel) => /[/.#$[\]]/.test(String(seedModel));
